extension-impl/registry-consul/src/main/java/com/alipay/sofa/rpc/registry/consul/ConsulRegistry.java

```diff
@@ -79,111 +95,46 @@
-     * 注册中心配置
-     *
-     * @param registryConfig 注册中心配置
-     */
     protected ConsulRegistry(RegistryConfig registryConfig) {
         super(registryConfig);
-    }
-
-    public String[] validateIp(RegistryConfig registryConfig) {
-        String addressInput = registryConfig.getAddress(); // xxx:2181,yyy:2181/path1/paht2
-
-        if (StringUtils.isEmpty(addressInput)) {
-            throw new SofaRpcRuntimeException("Address of consul registry is empty.");
-        }
-
-        int idx = addressInput.indexOf(CONTEXT_SEP);
-        String address; // IP地址
-        if (idx > 0) {
-            address = addressInput.substring(0, idx);
-            rootPath = addressInput.substring(idx);
-
-        } else {
-            address = addressInput;
-            rootPath = "/";
-        }
-
-        if (!ConsulURLUtils.isValidAddress(address)) {
-            throw new SofaRpcRuntimeException("Address format of consul registry is wrong.");
-        }
-        if (!rootPath.endsWith(CONTEXT_SEP)) {
-            rootPath += CONTEXT_SEP; // 保证以"/"结尾
-        }
-        String[] ipAndHost = StringUtils.split(address, ":");
-        return ipAndHost;
-    }
-
-    private ConsulService buildConsulHealthService(ConsulURL url) {
-        return ConsulService.newService()//
-            .withAddress(url.getHost())//
-            .withPort(Integer.toString(url.getPort()))//
-            .withName(ConsulURLUtils.toServiceName(url.getGroup()))//
-            .withTag(ConsulURLUtils.healthServicePath(url, ThrallRoleType.PROVIDER))//
-            .withId(url.getHost() + ":" + url.getPort() + "-" + url.getPath() + "-" + url.getVersion())//
-            .withCheckInterval(Integer.toString(ConsulConstants.TTL)).build();
-    }
-
-    private ConsulEphemeralNode buildEphemralNode(ConsulURL url, ThrallRoleType roleType) {
-        return ConsulEphemeralNode.newEphemralNode().withUrl(url)//
-            .withEphemralType(roleType)//
-            .withCheckInterval(Integer.toString(ConsulConstants.TTL * 6))//
-            .build();
+        this.properties = new ConsulRegistryProperties(registryConfig.getParameters());
     }
 
     @Override
     public void init() {
-
-        if (consulManager != null) {
+        if (consulClient != null) {
             return;
         }
 
-        String[] address = validateIp(registryConfig);
-        consulManager = new ConsulManager(address[0], Integer.parseInt(address[1]));
-        serviceCache = CacheBuilder.newBuilder().maximumSize(1000).build();
-        notifyExecutor = Executors.newCachedThreadPool(
-            new NamedThreadFactory("NotifyConsumerListener", true));
+        String[] hostAndPort = StringUtils.split(registryConfig.getAddress(), ":");
+        consulClient = new ConsulClient(hostAndPort[0], Integer.parseInt(hostAndPort[1]));
+
+        Integer coreSize = properties.getHeartbeatCoreSize();
+        if (coreSize == null) {
+            heartbeatExecutor = Executors.newSingleThreadScheduledExecutor();
```

Comment:
`newSingleThreadScheduledExecutor` creates a thread pool with only one thread, why is it dangerous?

Reply:
Just as [Alibaba-Java-Coding-Guidelines](https://alibaba.github.io/Alibaba-Java-Coding-Guidelines/) says:
> 4. [Mandatory] A thread pool should be created by ThreadPoolExecutor rather than Executors. These would make the parameters of the thread pool understandable. It would also reduce the risk of running out of system resource.
>>Note:
Below are the problems created by usage of Executors for thread pool creation: 
  1) FixedThreadPool and SingleThreadPool:
  Maximum request queue size Integer.MAX_VALUE. A large number of requests might cause OOM. 
  2) CachedThreadPool and ScheduledThreadPool:
  The number of threads which are allowed to be created is Integer.MAX_VALUE. Creating too many threads might lead to OOM.

For this usage, only use thread pool for schedule seems ok, but we can more rigorous.

Reply:
I see, i will use `ThreadPoolExecutor` instead and set a fixed queue size.